Randomization in tests

Right now all of our tests are using random data. This sometimes causes github actions to fail in unexpected ways.

We should use all deterministic data in our tests rather than random so that we get consistency across runs.

Example:

```
        for max_depth in [1, 3, 8, 10, 12, None]:
            model = model_type(max_depth=max_depth, **kwargs)
            X = np.random.rand(100, 200)
            X = np.array(X, dtype=np.float32)
            y = np.random.randint(num_classes, size=100)
    
            model.fit(X, y)
            pytorch_model = hummingbird.ml.convert(model, "pytorch", extra_config=extra_config)
            self.assertIsNotNone(pytorch_model)
>           np.testing.assert_allclose(model.predict(X), pytorch_model.predict(X), rtol=1e-06, atol=1e-06)
E           AssertionError: 
E           Not equal to tolerance rtol=1e-06, atol=1e-06
E           
E           Mismatched elements: 1 / 100 (1%)
E           Max absolute difference: 13.8999939
E           Max relative difference: 0.05819549
E            x: array([ 50.      , 307.      , 334.      , 707.341667, 422.      ,
E                   66.8     , 895.55    , 887.05    , 442.      , 969.383333,
E                  111.633333, 492.      , 419.116667, 501.075   , 718.633333,...
E            y: array([ 50.      , 307.      , 334.      , 707.3417  , 422.      ,
E                   66.8     , 895.55    , 887.05    , 442.      , 969.3833  ,
E                  111.63332 , 492.      , 419.1167  , 501.075   , 718.6333  ,...

tests/test_sklearn_decision_tree_converters.py:106: AssertionError
=========================== short test summary info ============================
FAILED tests/test_sklearn_decision_tree_converters.py::TestSklearnTreeConverter::test_extra_trees_gemm_regressor_converter
======================== 1 failed, 86 passed in 28.99s =========================
##[error]Process completed with exit code 1.
 Coverage
```

--- a/tests/test_backends.py
+++ b/tests/test_backends.py
@@ -18,6 +18,7 @@ class TestBackends(unittest.TestCase):
         max_depth = 10
         num_classes = 2
         model = GradientBoostingClassifier(n_estimators=10, max_depth=max_depth)
+        np.random.seed(0)
         X = np.random.rand(100, 200)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(num_classes, size=100)
@@ -34,6 +35,7 @@ class TestBackends(unittest.TestCase):
         max_depth = 10
         num_classes = 2
         model = GradientBoostingClassifier(n_estimators=10, max_depth=max_depth)
+        np.random.seed(0)
         X = np.random.rand(100, 200)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(num_classes, size=100)

--- a/tests/test_lightgbm_converters.py
+++ b/tests/test_lightgbm_converters.py
@@ -15,6 +15,7 @@ class TestLGBMConverter(unittest.TestCase):
     # Check tree implementation
     def test_lgbm_implementation(self):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.random.rand(10, 1)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(2, size=10)
@@ -35,6 +36,7 @@ class TestLGBMConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12, None]:
             model = lgb.LGBMClassifier(n_estimators=10, max_depth=max_depth)
+            np.random.seed(0)
             X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(num_classes, size=100)
@@ -81,6 +83,7 @@ class TestLGBMConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12, None]:
             model = lgb.LGBMRegressor(n_estimators=10, max_depth=max_depth)
+            np.random.seed(0)
             X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(num_classes, size=100)

--- a/tests/test_sklearn_decision_tree_converters.py
+++ b/tests/test_sklearn_decision_tree_converters.py
@@ -17,6 +17,7 @@ class TestSklearnTreeConverter(unittest.TestCase):
     # Check tree implementation
     def test_random_forest_implementation(self):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.random.rand(1, 1)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(2, size=1)
@@ -37,6 +38,7 @@ class TestSklearnTreeConverter(unittest.TestCase):
     def _run_tree_classification_converter(self, model_type, num_classes, extra_config={}, labels_shift=0, **kwargs):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12, None]:
+            np.random.seed(0)
             X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(num_classes, size=100) + labels_shift
@@ -53,15 +55,21 @@ class TestSklearnTreeConverter(unittest.TestCase):
 
     # Random forest gemm classifier
     def test_random_forest_gemm_classifier_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 2, extra_config={"tree_implementation": "gemm"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 2, extra_config={"tree_implementation": "gemm"}, n_estimators=10
+        )
 
     # Random forest tree_trav classifier
     def test_random_forest_tree_trav_classifier_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 2, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 2, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10
+        )
 
     # Random forest perf_tree_trav classifier
     def test_random_forest_perf_tree_trav_classifier_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 2, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 2, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10
+        )
 
     # Random forest multi classifier
     def test_random_forest_multi_classifier_converter(self):
@@ -69,33 +77,46 @@ class TestSklearnTreeConverter(unittest.TestCase):
 
     # Random forest gemm multi classifier
     def test_random_forest_gemm_multi_classifier_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 3, extra_config={"tree_implementation": "gemm"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 3, extra_config={"tree_implementation": "gemm"}, n_estimators=10
+        )
 
     # Random forest tree_trav multi classifier
     def test_random_forest_tree_trav_multi_classifier_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 3, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 3, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10
+        )
 
     # Random forest perf_tree_trav multi classifier
     def test_random_forest_perf_tree_trav_multi_classifier_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 3, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 3, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10
+        )
 
     # Random forest gemm classifier shifted classes
     def test_random_forest_gemm_classifier_shifted_labels_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 3, labels_shift=2, extra_config={"tree_implementation": "gemm"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 3, labels_shift=2, extra_config={"tree_implementation": "gemm"}, n_estimators=10
+        )
 
     # Random forest tree_trav classifier shifted classes
     def test_random_forest_tree_trav_classifier_shifted_labels_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 3, labels_shift=2, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 3, labels_shift=2, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10
+        )
 
     # Random forest perf_tree_trav classifier shifted classes
     def test_random_forest_perf_tree_trav_classifier_shifted_labels_converter(self):
-        self._run_tree_classification_converter(RandomForestClassifier, 3, labels_shift=2, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10)
+        self._run_tree_classification_converter(
+            RandomForestClassifier, 3, labels_shift=2, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10
+        )
 
     # Used for regression tests
     def _run_tree_regressor_converter(self, model_type, num_classes, extra_config={}, **kwargs):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12, None]:
             model = model_type(max_depth=max_depth, **kwargs)
+            np.random.seed(0)
             X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(num_classes, size=100)
@@ -111,15 +132,21 @@ class TestSklearnTreeConverter(unittest.TestCase):
 
     # Random forest gemm regressor
     def test_random_forest_gemm_regressor_converter(self):
-        self._run_tree_regressor_converter(RandomForestRegressor, 1000, extra_config={"tree_implementation": "gemm"}, n_estimators=10)
+        self._run_tree_regressor_converter(
+            RandomForestRegressor, 1000, extra_config={"tree_implementation": "gemm"}, n_estimators=10
+        )
 
     # Random forest tree_trav regressor
     def test_random_forest_tree_trav_regressor_converter(self):
-        self._run_tree_regressor_converter(RandomForestRegressor, 1000, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10)
+        self._run_tree_regressor_converter(
+            RandomForestRegressor, 1000, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10
+        )
 
     # Random forest perf_tree_trav regressor
     def test_random_forest_perf_tree_trav_regressor_converter(self):
-        self._run_tree_regressor_converter(RandomForestRegressor, 1000, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10)
+        self._run_tree_regressor_converter(
+            RandomForestRegressor, 1000, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10
+        )
 
     # Extra trees regressor
     def test_extra_trees_regressor_converter(self):
@@ -127,15 +154,21 @@ class TestSklearnTreeConverter(unittest.TestCase):
 
     # Extra trees gemm regressor
     def test_extra_trees_gemm_regressor_converter(self):
-        self._run_tree_regressor_converter(ExtraTreesRegressor, 1000, extra_config={"tree_implementation": "gemm"}, n_estimators=10)
+        self._run_tree_regressor_converter(
+            ExtraTreesRegressor, 1000, extra_config={"tree_implementation": "gemm"}, n_estimators=10
+        )
 
     # Extra trees tree_trav regressor
     def test_extra_trees_tree_trav_regressor_converter(self):
-        self._run_tree_regressor_converter(ExtraTreesRegressor, 1000, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10)
+        self._run_tree_regressor_converter(
+            ExtraTreesRegressor, 1000, extra_config={"tree_implementation": "tree_trav"}, n_estimators=10
+        )
 
     # Extra trees perf_tree_trav regressor
     def test_extra_trees_perf_tree_trav_regressor_converter(self):
-        self._run_tree_regressor_converter(ExtraTreesRegressor, 1000, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10)
+        self._run_tree_regressor_converter(
+            ExtraTreesRegressor, 1000, extra_config={"tree_implementation": "perf_tree_trav"}, n_estimators=10
+        )
 
     # Decision tree regressor
     def test_decision_tree_regressor_converter(self):
@@ -164,6 +197,7 @@ class TestSklearnTreeConverter(unittest.TestCase):
     # Used for small tree tests
     def _run_random_forest_classifier_single_node_tree_converter(self, extra_config={}):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.random.rand(1, 1)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(1, size=1)
@@ -187,6 +221,7 @@ class TestSklearnTreeConverter(unittest.TestCase):
     # Failure Cases
     def test_random_forest_classifier_raises_wrong_type(self):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.random.rand(100, 200)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(3, size=100).astype(np.float32)  # y must be int, not float, should error
@@ -195,6 +230,7 @@ class TestSklearnTreeConverter(unittest.TestCase):
 
     def test_random_forest_classifier_raises_wrong_extra_config(self):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.array(np.random.rand(100, 200), dtype=np.float32)
         y = np.random.randint(3, size=100)
         model = RandomForestClassifier(n_estimators=10).fit(X, y)

--- a/tests/test_sklearn_gbdt_converters.py
+++ b/tests/test_sklearn_gbdt_converters.py
@@ -16,6 +16,7 @@ class TestSklearnGradientBoostingConverter(unittest.TestCase):
     # Check tree implementation
     def test_gbdt_implementation(self):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.random.rand(10, 1)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(2, size=10)
@@ -40,6 +41,7 @@ class TestSklearnGradientBoostingConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12, None]:
             model = GradientBoostingClassifier(n_estimators=10, max_depth=max_depth)
+            np.random.seed(0)
             X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(num_classes, size=100) + labels_shift
@@ -53,6 +55,7 @@ class TestSklearnGradientBoostingConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12, None]:
             model = GradientBoostingRegressor(n_estimators=10, max_depth=max_depth)
+            np.random.seed(0)
             X = np.random.rand(100, 200).astype(np.float32)
             y = np.random.normal(size=100)
 
@@ -119,6 +122,7 @@ class TestSklearnGradientBoostingConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12, None]:
             model = GradientBoostingClassifier(n_estimators=10, max_depth=max_depth, init="zero")
+            np.random.seed(0)
             X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(3, size=100)
@@ -132,6 +136,7 @@ class TestSklearnGradientBoostingConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12, None]:
             model = GradientBoostingRegressor(n_estimators=10, max_depth=max_depth, init="zero")
+            np.random.seed(0)
             X = np.random.rand(100, 200).astype(np.float32)
             y = np.random.normal(size=100)
 
@@ -143,6 +148,7 @@ class TestSklearnGradientBoostingConverter(unittest.TestCase):
     # Failure Cases
     def test_sklearn_GBDT_classifier_raises_wrong_type(self):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.random.rand(100, 200)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(3, size=100).astype(np.float32)  # y must be int, not float, should error

--- a/tests/test_sklearn_histgbdt_converters.py
+++ b/tests/test_sklearn_histgbdt_converters.py
@@ -17,9 +17,10 @@ class TestSklearnHistGradientBoostingClassifier(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [2, 3, 8, 10, 12, None]:
             model = HistGradientBoostingClassifier(max_iter=10, max_depth=max_depth)
-            X = [[0, 1], [1, 1], [2, 0]]
+            np.random.seed(0)
+            X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
-            y = np.array([100, -10, 50]) + labels_shift
+            y = np.random.randint(num_classes, size=100) + labels_shift
 
             model.fit(X, y)
             pytorch_model = hummingbird.ml.convert(model, "pytorch", extra_config=extra_config)
@@ -67,6 +68,7 @@ class TestSklearnHistGradientBoostingClassifier(unittest.TestCase):
     # Failure Cases
     def test_HistGBDT_raises_wrong_type(self):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.random.rand(100, 200)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(3, size=100).astype(np.float32)  # y must be int, not float, should error

--- a/tests/test_xgboost_converters.py
+++ b/tests/test_xgboost_converters.py
@@ -15,6 +15,7 @@ class TestXGBoostConverter(unittest.TestCase):
     # Check tree implementation
     def test_xgb_implementation(self):
         warnings.filterwarnings("ignore")
+        np.random.seed(0)
         X = np.random.rand(1, 1)
         X = np.array(X, dtype=np.float32)
         y = np.random.randint(2, size=1)
@@ -35,6 +36,7 @@ class TestXGBoostConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12]:
             model = xgb.XGBClassifier(n_estimators=10, max_depth=max_depth)
+            np.random.seed(0)
             X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(num_classes, size=100)
@@ -81,6 +83,7 @@ class TestXGBoostConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for max_depth in [1, 3, 8, 10, 12]:
             model = xgb.XGBRegressor(n_estimators=10, max_depth=max_depth)
+            np.random.seed(0)
             X = np.random.rand(100, 200)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(num_classes, size=100)
@@ -111,6 +114,7 @@ class TestXGBoostConverter(unittest.TestCase):
         warnings.filterwarnings("ignore")
         for extra_config_param in ["tree_trav", "perf_tree_trav", "gemm"]:
             model = xgb.XGBClassifier(n_estimators=1, max_depth=1)
+            np.random.seed(0)
             X = np.random.rand(1, 1)
             X = np.array(X, dtype=np.float32)
             y = np.random.randint(2, size=1)

